src/logic/hosts.rs

```diff
@@ -0,0 +1,69 @@
+use std::collections::HashSet;
+
+use chrono::Utc;
+use thiserror::Error;
+
+use crate::db::{
+    models::{Host, HostId, LeasedHost, UserId},
+    Registry,
+};
+
+#[derive(Error, Debug)]
+pub enum HostError {
+    #[error("Database error")]
+    DatabaseError(#[from] sqlx::Error),
+
+    #[error("Host is already leased")]
+    AlreadyLeased(Vec<HostId>),
+}
+
+pub async fn get_available_hosts(registry: &Registry) -> Result<Vec<Host>, HostError> {
+    let mut tx = registry.begin().await?;
+    let hosts = tx.get_available_hosts().await?;
+    tx.commit().await?;
+    Ok(hosts)
+}
+
+pub async fn lease(
+    registry: &Registry,
+    user_id: &UserId,
+    hosts_ids: &[HostId],
+    lease_for: chrono::TimeDelta,
+) -> Result<Vec<LeasedHost>, HostError> {
+    let mut tx = registry.begin().await?;
+    let available: HashSet<_> = tx
+        .get_available_hosts()
+        .await?
+        .into_iter()
+        .map(|h| h.id)
+        .collect();
+    let mut hosts_ids_set: HashSet<_> = HashSet::new();
```

Comment:
```suggestion
    let mut hosts_ids_set: HashSet<_> = HashSet::from(hosts_ids.to_vec());
```
?

Reply:
the compiler is not happy with this:
`the trait bound `HashSet<_, _>: std::convert::From<Vec<HostId>>` is not satisfied`

Reply:
strange, it must be simpler...